src/sqs/error-handlers.ts

```diff
@@ -0,0 +1,69 @@
+import {Context, ErrorHandler, ErrorHandlingResult} from "./index";
+import {JiraClientError} from "../jira/client/axios";
+import {Octokit} from "@octokit/rest";
+import {RateLimitingError} from "../config/enhance-octokit";
+import {emitWebhookFailedMetrics} from "../util/webhooks";
+import {PushQueueMessagePayload} from "./push";
+
+/**
+ * Sometimes we can get errors from Jira and GitHub which does not indicate a failured webhook. For example:
+ *  Jira site is gone, we'll get 404
+ *  GitHub App is not installed anymore we'll get 401
+ *  etc.
+ *
+ *  In such cases webhook processing doesn't make sense anymore and we need to silently discard these errors
+ */
+const UNRETRYABLE_STATUS_CODES = [401, 404, 403];
+
+const RATE_LIMITING_DELAY_BUFFER_SEC = 10;
+const EXPONENTIAL_BACKOFF_BASE_SEC = 60;
+const EXPONENTIAL_BACKOFF_MULTIPLIER = 3;
+
+export const jiraOctokitErrorHandler : ErrorHandler<any> = async (error: JiraClientError | Octokit.HookError | RateLimitingError | Error,
+	context: Context<any>) : Promise<ErrorHandlingResult> => {
+
+	const maybeResult = maybeHandleNonFailureCase(error, context);
+	if (maybeResult) {
+		return maybeResult;
+	}
+
+	const errorHandlingResult = handleFailureCase(error, context);
+
+	if (!errorHandlingResult.retryable || context.lastAttempt ) {
+		context.log.error({error}, "Webhook push processing failed and won't be retried anymore");
+		emitWebhookFailedMetrics("push")
+	}
+
+	return errorHandlingResult;
+}
+
+function maybeHandleNonFailureCase(error: Error, context: Context<PushQueueMessagePayload>): ErrorHandlingResult | undefined {
+	if (error instanceof JiraClientError &&
+		error.status &&
+		UNRETRYABLE_STATUS_CODES.includes(error.status)) {
+		context.log.warn(`Received ${error.status} from Jira. Unretryable. Discarding the message`);
+		return {retryable: false}
+	}
+
+	//If error is Octokit.HookError, then we need to check the response status
+	//Unfortunately we can't check if error is instance of Octokit.HookError because it is not a calss, so we'll just rely on status
+	//TODO Add error handling for the new GitHub client when it will be done
+	const maybeErrorWithStatus : any = error;
```

Comment:
This variable name makes my brain hurt. Maybe something like `statusHasError` instead?

Reply:
I copy-pasted the content 1:1, don't want to modify it to keep the scope limited. Can fix in the following PR ;)